### DISCUSSION TITLE
Nice GRPC web 🚀

### DISCUSSION THREAD

**Discussion by ChristopheBougere** [deeplay-io/nice-grpc#9] [General]
I just saw that you refactored the repo with lerna to enable a nice-grpc-web package 🤩

As nice-grpc already made my code a lot cleaner, and as I'll have to implement a web version soonish, I am quite curious:
Are you planning to wrap the official [grpc/grpc-web](https://github.com/grpc/grpc-web)? Or maybe [improbable-eng/grpc-web](https://github.com/improbable-eng/grpc-web)?

Can't wait to see this released 🤗

**Comment by aikoven** [comment]
Hey, thank you for your interest.

I've been working on it for the past week, and I can say that it's mostly finished. You can expect the release somewhere next week.

What's left are the docs (mostly repeating the docs for `nice-grpc`, so should not take long) and some internal testing with my company's web apps. 

Also one missing part is the support for `ts-proto` (waiting for PR https://github.com/stephenh/ts-proto/pull/316 to get merged).

`nice-grpc-web` will be based on `@improbable-eng/grpc-web`. I chose it because of the support for client- and bidi-streaming calls via their custom websocket proxy. Though it also works fine with official Envoy filter, as well as Contour ingress controller for Kubernetes, which is based on Envoy. With Envoy, only unary and server streaming methods are supported, although I find it more convenient to use.

There will be another package, `nice-grpc-common` containing common data structures and types for both `nice-grpc` and `nice-grpc-web`, like `Metadata` and middleware types. This would allow to build isomorphic client middleware that can be used on both Node.js and the browser. This would introduce small breaking changes to `nice-grpc`.

**Comment by ChristopheBougere** [comment]
Oh that's great to know, thanks for all the details 👌

I was also leaning towards `@improbable-eng/grpc-web` for the client-side support, but my choice wasn't 100% made.

Also, I didn't know about `ts-proto`, but I may have a look as I have some limitations with `google-protobuf` for enums.

**Comment by aikoven** [comment]
I published all the fresh stuff — check it out!

I highly recommend `ts-proto`, it's much more ergonomic and feature-rich than `google-protobuf`, and it's also very customizable via many different options.

**Reply by ChristopheBougere** [reply]
Thanks, I gave it a first try (along with `ts-proto`) and it seems to work pretty well 👌

**Reply by ChristopheBougere** [reply]
Also, about `ts-proto`, I needed to add the following option to the `grpc_tools_node_protoc` command to make it work: `--plugin=./node_modules/.bin/protoc-gen-ts_proto`
Not sure if that's specific to my setup

**Reply by aikoven** [reply]
Plugins are auto-discovered when they are in the `PATH`. See https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.compiler.plugin

If you place compilation command into a `package.json` script, NPM includes `node_modules/.bin` in the `PATH` automatically. 

I'll update the docs to mention this.

**Reply by ChristopheBougere** [reply]
Ah OK, good to know 👍
